libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -835,67 +842,509 @@ bool GCS_MAVLINK::handle_mission_item(mavlink_message_t *msg, AP_Mission &missio
+void GCS_MAVLINK::find_next_bucket_to_send()
+{
+#if DEBUG_SEND_MESSAGE_TIMINGS
+    void *data = hal.scheduler->disable_interrupts_save();
+    uint32_t start_us = AP_HAL::micros();
+#endif
+
+    const uint16_t now16_ms{AP_HAL::millis16()};
+
+    // all done sending this bucket... find another bucket...
+    sending_bucket_id = no_bucket_to_send;
+    uint16_t ms_before_send_next_bucket_to_send = UINT16_MAX;
+    for (uint8_t i=0; i<ARRAY_SIZE(deferred_message_bucket); i++) {
+        if (deferred_message_bucket[i].ap_message_ids.first_set() == -1) {
+            // no entries
+            continue;
+        }
+        const uint16_t interval = reschedule_interval(deferred_message_bucket[i]);
+        const uint16_t ms_since_last_sent = now16_ms - deferred_message_bucket[i].last_sent_ms;
+        uint16_t ms_before_send_this_bucket;
+        if (ms_since_last_sent > interval) {
+            // should already have sent this bucket!
+            ms_before_send_this_bucket = 0;
+        } else {
+            ms_before_send_this_bucket = interval - ms_since_last_sent;
+        }
+        if (sending_bucket_id == no_bucket_to_send ||
+            ms_before_send_this_bucket < ms_before_send_next_bucket_to_send) {
+            sending_bucket_id = i;
+            ms_before_send_next_bucket_to_send = ms_before_send_this_bucket;
         }
-        num_deferred_messages--;
     }
+    if (sending_bucket_id != no_bucket_to_send) {
+        bucket_message_ids_to_send = deferred_message_bucket[sending_bucket_id].ap_message_ids;
+    } else {
+        bucket_message_ids_to_send.clearall();
+    }
+
+#if DEBUG_SEND_MESSAGE_TIMINGS
+    uint32_t delta_us = AP_HAL::micros() - start_us;
+    hal.scheduler->restore_interrupts(data);
+    if (delta_us > try_send_message_stats.fnbts_maxtime) {
+        try_send_message_stats.fnbts_maxtime = delta_us;
+    }
+#endif
 }
 
-void GCS_MAVLINK::retry_deferred()
+ap_message GCS_MAVLINK::next_deferred_message_to_send()
 {
-    push_deferred_messages();
+    if (sending_bucket_id == no_bucket_to_send) {
+        // could happen if all streamrates are zero?
+        return no_message_to_send;
+    }
+
+    const uint16_t now16_ms = AP_HAL::millis16();
+    const uint16_t ms_since_last_sent = now16_ms - deferred_message_bucket[sending_bucket_id].last_sent_ms;
+    if (ms_since_last_sent < deferred_message_bucket[sending_bucket_id].interval_ms) {
+        // not time to send this bucket
+        return no_message_to_send;
+    }
+
+    const int16_t next = bucket_message_ids_to_send.first_set();
+    if (next == -1) {
+        // should not happen
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+        AP_HAL::panic("next_deferred_message_to_send called on empty bucket");
+#endif
+        find_next_bucket_to_send();
+    }
+    return (ap_message)next;
 }
 
-// send a message using mavlink, handling message queueing
-void GCS_MAVLINK::send_message(enum ap_message id)
+bool GCS_MAVLINK::do_try_send_message(const ap_message id)
```

Comment:
could we add a comment to clarify how "do_try_send_message" is different from "try_send_message"?

Reply:
I've added a comment.  It's just a wrapper which is not expected to be overridden by the subclasses.  Obviously my imagination for method naming is lacking here :-)